enhancements/machine-api/machine-health-check-pause-on-upgrade.md

```diff
@@ -0,0 +1,175 @@
+---
+title: machine-health-check-pause-on-cluster-upgrade
+
+authors:
+  - @slintes
+
+reviewers:
+  - @JoelSpeed
+  - @michaelgugino
+  - @beekhof
+
+approvers:
+  - @JoelSpeed
+  - @michaelgugino
+  - @beekhof
+
+creation-date: 2021-07-14  
+last-updated: 2021-07-14  
+status: implementable  
+
+see-also:
+  - "enhancements/machine-api/machine-health-checking.md"
+---
+
+# Pause MachineHealthChecks During Cluster Upgrades
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Automatically prevent unwanted remediation during cluster upgrades caused by temporarily unhealthy nodes.
+
+## Motivation
+
+During cluster upgrades nodes get temporarily unhealthy, which might trigger remediation when a machineHealthCheck
+resource is targeting those nodes, and they don't get healthy before the configured timeout. This remediation delays
+the cluster upgrade significantly especially on bare metal clusters, and causes needless interruption of customer
+workloads while they are transferred to other nodes.
+
+### Goals
+
+- Automatically pause all machineHealthChecks during cluster upgrades
```

Comment:
Let's try walking before we attempt to run

Reply:
It doesn't contradicts.
it helps to know what is the vision/end-goal so we can:

1. know what are the limitations of the proposed solution
2. Make steps at the right direction

Reply:
added this to non-goals and alternatives

Reply:
I think you actually care about the machineConfigPool being in progressing state rather than the upgrade state. A full hour of the upgrade has nothing to do with host reboots and if we deliver on paused-worker-pool upgrades most nodes in a cluster won't even reboot some of the time.

I think you should instead watch for nodes that have been cordoned and ignore them in some manner. That seems more broadly applicable.

@JoelSpeed do you have an opinion on this?

Reply:
> I think you actually care about the machineConfigPool being in progressing state rather than the upgrade state.

I agree, if we can track the MCP upgrade rather than whole cluster upgrade, that will be a smaller time window. Perhaps the easiest way to do this is to actually track the `progressing` state of the ClusterOperator for machine config?

> I think you should instead watch for nodes that have been cordoned and ignore them in some manner. That seems more broadly applicable.

This seems more like the node maintenance lease idea that has floated around before, https://github.com/kubernetes/enhancements/pull/1411, this is definitely something that could be more intuitive and avoid the dependency on upgrade specific watching, but is a much larger piece of work.

In terms of just watching cordoned nodes, to me this makes sense as a signal that a node is going into maintenance of some description. Need to think a bit more about if this would cause other issues 🤔